lib/std/Build/Step/Compile.zig

```diff
@@ -1077,9 +1077,8 @@ fn make(step: *Step, prog_node: *std.Progress.Node) !void {
                             .exe => return step.fail("cannot link with an executable build artifact", .{}),
                             .@"test" => return step.fail("cannot link with a test", .{}),
                             .obj => {
-                                const included_in_lib = !my_responsibility and
-                                    compile.kind == .lib and other.kind == .obj;
-                                if (!already_linked and !included_in_lib) {
+                                const included_in_lib_or_obj = !my_responsibility and (compile.kind == .lib or compile.kind == .obj);
+                                if (!already_linked and !included_in_lib_or_obj) {
```

Comment:
I tested these changes against [contrib-testing](https://github.com/ziglang/contrib-testing/) and they passed :+1: